src/snowflake/cli/api/metrics.py

```diff
@@ -54,17 +67,125 @@ class CLICounterField:
     PACKAGE_SCRIPTS = f"{_TypePrefix.FEATURES}.{_DomainPrefix.APP}.package_scripts"
 
 
+class _CLIMetricsStep:
+    """
+    class for holding metrics step data and encapsulating related operations
+    """
+
+    _id_counter = count(start=1, step=1)
+
+    # keys for dict representation
+    ID_KEY = "id"
+    NAME_KEY = "name"
+    PARENT_KEY = "parent"
+    PARENT_ID_KEY = "parent_id"
+    START_TIME_KEY = "start_time"
+    EXECUTION_TIME_KEY = "execution_time"
+    ERROR_KEY = "error"
+
+    def __init__(self, name: str, parent: _CLIMetricsStep | None = None):
+        if not name:
+            raise CLIMetricsInvalidUsageError("step name must not be empty")
+
+        # to disambiguate steps with same name
+        self._step_id: int = next(self._id_counter)
+        self._name: str = name
+
+        self._parent: Optional[str] = parent.name if parent is not None else None
+        self._parent_id: Optional[int] = parent.step_id if parent is not None else None
```

Comment:
nit: not a huge fan of expanding things this way. Why can't we just keep a ref to the parent and read the data on the fly as needed?